src/packet/gre.rs.in

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2016 Robert Collins <robertc@robertcollins.net>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#[cfg(test)]
+use packet::Packet;
+
+use packet::PrimitiveValues;
+
+use pnet_macros_support::types::*;
+
+/// GRE Packet.
+///
+/// Note that routing information from RFC 1701 is not implemented, packets
+/// with routing_present true will have incorrect payload lengths.
```

Comment:
See rfc 1701 page 4 - the SRE structure is basically a null terminated list of variable length structures. So implementing this efficiently is a mess - something like:
length = if 0 == routing_present { 0 } else { while true length += parse_sre...  break }

I'd be delighted to put a guard in to ensure we don't mishandle a packet with this set. For now, I'll throw a panic in :). At least it won't be a silent failure.